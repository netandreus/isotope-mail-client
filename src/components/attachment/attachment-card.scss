@import '../../styles/partials/variables';

.attachment {
  position: relative;
  display: flex;
  align-items: center;

  background-color: darken($base-background-color, 3%);
  margin: 0 6px 6px 0;
  padding: 6px;
  max-width: 200px;

  cursor: pointer;

  &:hover {
    background-color: darken($base-background-color, 10%);
  }

  &.downloading {
    cursor: initial;
  }

  .spinner {
    position: absolute;
    top: 0px;
    left: 0px;
    height: 100%;
    width: 100%;
    background-color: rgba(darken($base-background-color, 20%), 0.6);

    .canvas {
      height: 22px;
    }
  }

  .icon {
    margin-right: 6px;
  }

  .fileInfo {
    overflow: hidden;

    .fileName {
      display: block;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;

      font-size: 0.90 * $font-size-body;
      color: lighten($base-color, 5%);
    }

    .size {
      display: block;
      font-size: 0.70 * $font-size-body;

      color: lighten($base-color, 15%);
    }
  }
}
