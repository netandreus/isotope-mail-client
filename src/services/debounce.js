/**
 * Debounce a function so that it get called only once in (at minimum) the specified period
 *
 * @param func to be debounced
 * @param period in milliseconds
 * @returns {Function} The debounced function
 */
function debounce(func, period = 100) {
  let funcTimeout;
  function debounced() {
    const funcContext = this;
    const funcArguments = arguments;
    const runDebounced = () => {
      funcTimeout = null;
      func.apply(funcContext, funcArguments);
    };
    clearTimeout(funcTimeout);
    funcTimeout = setTimeout(runDebounced, period);
  }
  debounced.cancel = function () {
    clearTimeout(funcTimeout);
  };
  return debounced;
}


export default debounce;
