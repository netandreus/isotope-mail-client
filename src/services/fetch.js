import {refreshUserCredentials} from '../actions/application';

export const HttpHeaders = {
  ISOTOPE_CREDENTIALS: 'Isotope-Credentials',
  ISOTOPE_SALT: 'Isotope-Salt',
  CONTENT_TYPE: 'Content-Type'
};

export const HttpStatusFamilies = {
  INFORMATIONAL: 1,
  SUCCESSFUL: 2,
  REDIRECTION: 3,
  CLIENT_ERROR: 4,
  SERVER_ERROR: 5
};

export class AuthenticationException extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = 'AuthenticationException';
  }
}

/**
 * Object to store the different AbortController(s) that will be used in the service methods to fetch from the API backend.
 *
 * @type {Object}
 * @private
 */
export const abortControllerWrappers = {};

/**
 * Converts response Promise to json Promise or throws an error.
 *
 * @param {Response}response
 * @returns {PromiseLike<object>}
 */
export function toJson(response) {
  if (!response.ok) {
    if (response.status === 401 || response.status === 403) {
      throw new AuthenticationException(response.statusText);
    } else {
      throw Error(`${response.status}`);
    }
  }
  return response.json();
}

/**
 * Dispatches the refreshUserCredentials action with the values obtained from the Http Response headers.
 *
 * @param dispatch
 * @param response
 * @returns {*}
 */
export function refreshCredentials(dispatch, response) {
  const credentials = response.headers.get(HttpHeaders.ISOTOPE_CREDENTIALS);
  const salt = response.headers.get(HttpHeaders.ISOTOPE_SALT);
  if (credentials && salt) {
    dispatch(refreshUserCredentials(credentials, salt));
  }
  return Promise.resolve(response);
}

/**
 * Triggers the abort method of a the specified AbortController
 *
 * @param abortController {AbortController}
 */
export function abortFetch(abortController) {
  if (abortController && abortController.abort) {
    abortController.abort();
  }
}

/**
 * Creates a new headers object including Isotope specific credentials headers for server authentication.
 *
 * @param {{}} credentials
 * @param {{}} [originalHeaders={}] originalHeaders
 * @returns {{}}
 */
export function credentialsHeaders(credentials, originalHeaders) {
  const newHeaders = Object.assign({}, originalHeaders ? originalHeaders : {});
  newHeaders[HttpHeaders.ISOTOPE_CREDENTIALS] = credentials.encrypted;
  newHeaders[HttpHeaders.ISOTOPE_SALT] = credentials.salt;
  return newHeaders;
}

/**
 * Returns true if the response the Successful HTTP status code range (2xx)
 *
 * @param HTTP status code of the response
 * @returns {boolean} true if status code belongs to HttpStatusFamilies.SUCCESSFUL family
 */
export function isSuccessful(status) {
  return Math.floor(status / 100) === HttpStatusFamilies.SUCCESSFUL;
}

