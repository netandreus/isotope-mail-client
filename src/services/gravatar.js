import md5 from './md5';

const BASE_URL = 'https://www.gravatar.com';

export function imageUrl(email, options = {}) {
  const hash = md5(email.trim().toLowerCase());
  const url = new URL(`${BASE_URL}/avatar/${hash}`);
  if (options.defaultImage) {
    url.searchParams.append('d', options.defaultImage);
  }
  return url.toString();
}
