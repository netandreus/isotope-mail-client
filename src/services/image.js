
const dataUrlToBlob = dataUrl => {
  const dataUrlParts = dataUrl.split(',');
  const mime = dataUrlParts[0].match(/:(.*?);/)[1];
  const binStr = atob(dataUrlParts[1]);
  const len = binStr.length;
  const u8arr = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    u8arr[i] = binStr.charCodeAt(i);
  }
  return new Blob([u8arr], {type: mime});
};

export const compressImage = (imageBlob, type = 'image/jpeg', quality = 0.85) => new Promise(resolve => {
  const imageUrl = URL.createObjectURL(imageBlob);
  const image = new Image();
  image.onload = () => {
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');
    ctx.imageSmoothingQuality = 'high';
    ctx.imageSmoothingEnabled = true;
    ctx.mozImageSmoothingEnabled = true;
    ctx.oImageSmoothingEnabled = true;
    ctx.webkitImageSmoothingEnabled = true;
    canvas.width = image.width;
    canvas.height = image.height;
    ctx.drawImage(image, 0, 0, canvas.width, canvas.height);
    URL.revokeObjectURL(imageUrl);
    resolve(dataUrlToBlob(canvas.toDataURL(type, quality)));
  };
  image.src = imageUrl;
});

