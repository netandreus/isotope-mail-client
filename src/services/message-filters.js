
const getAll = messages => messages;
const getRead = messages => messages.filter(message => message.seen === true);
const getUnread = messages => messages.filter(message => message.seen === false);
const getFlagged = messages => messages.filter(message => message.flagged === true);

const MessageFilters = [
  {key: 'ALL', i18nKey: 'messageList.filters.All', selector: getAll},
  {key: 'READ', i18nKey: 'messageList.filters.Read', selector: getRead},
  {key: 'UNREAD', i18nKey: 'messageList.filters.Unread', selector: getUnread},
  {key: 'FLAGGED', i18nKey: 'messageList.filters.Flagged', selector: getFlagged}
].reduce((acc, mf) => {
  acc[mf.key] = mf;
  return acc;
}, {});

export const getFromKey = (key = null) => {
  if (key && MessageFilters[key]) {
    return MessageFilters[key];
  }
  return MessageFilters.ALL;
};

export default MessageFilters;
