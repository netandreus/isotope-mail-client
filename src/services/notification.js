import i18n from './i18n';

const GRANTED_PERMISSION = 'granted';
const DENIED_PERMISSION = 'denied';

function requestPermission() {
  // Request only if necessary
  if ([GRANTED_PERMISSION, DENIED_PERMISSION].includes(Notification.permission)) {
    return Promise.resolve(Notification.permission);
  }
  return Notification.requestPermission();
}

export function notify(message, options) {
  return requestPermission().then(permission => {
    if (permission === GRANTED_PERMISSION) {
      const notification = new Notification(message, {...options, icon: 'assets/images/notification-icon.png'});
      notification.onclick = () => window.focus();
      return notification;
    }
    return null;
  });
}

export function notifyNewMail() {
  const t = i18n.t.bind(i18n);
  return notify(t('notifications.newMail'), {tag: 'isotope.new-mail', renotify: true});
}
