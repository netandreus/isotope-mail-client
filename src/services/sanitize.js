import sanitize from 'dompurify';
import ReactDOMServer from 'react-dom/server';
import {renderSpinner} from '../components/spinner/spinner';

sanitize.addHook('afterSanitizeAttributes', node => {
  // set all elements owning target to target=_blank -> Open URLS in new tab
  if ('target' in node) {
    node.setAttribute('target', '_blank');
  }
  // set non-HTML/MathML links to xlink:show=new
  if (!node.hasAttribute('target')
    && (node.hasAttribute('xlink:href')
      || node.hasAttribute('href'))) {
    node.setAttribute('xlink:show', 'new');
  }

  // FIX Image aspect ratio
  if (node.nodeName === 'IMG' && node.width) {
    node.style.width = '';
    node.style.height = '';
    if (node.width) {
      node.style.width = `${node.width}px`;
      node.removeAttribute('width');
    }
    node.removeAttribute('height');
  }

  // FIX HR overflow
  if (node.nodeName === 'HR' && node.width) {
    node.removeAttribute('width');
  }
});

const ISOTOPE_WHITELISTED_URL = '#IsotopeEmbedded';

// 1/2 Replaces "insecure" blob: uris before sanitation for controlled ones
sanitize.addHook('uponSanitizeAttribute', (node, hookEvent) => {
  if (node.nodeName === 'IMG' && hookEvent.attrName === 'src' && hookEvent.attrValue.indexOf('blob:') === 0) {
    node.src = `${node.src.substring(5)}${ISOTOPE_WHITELISTED_URL}`;
    hookEvent.attrValue = node.src;
  }
});

// 2/2 Replaces previously parsed "insecure" urls to original ones -> Whitelist effect
// Hides cid: images that are being loaded
sanitize.addHook('afterSanitizeAttributes', node => {
  if (node.nodeName === 'IMG' && node.src && node.src.indexOf(ISOTOPE_WHITELISTED_URL) > -1) {
    node.src = `blob:${node.src.replace(ISOTOPE_WHITELISTED_URL, '')}`;
  }
  if (node.nodeName === 'IMG' && node.src && node.src.indexOf('cid:') === 0) {
    const spinner = document.createElement('div');
    spinner.innerHTML = ReactDOMServer.renderToStaticMarkup(renderSpinner());
    spinner.getElementsByClassName('canvas')[0].style.height = node.height ? node.height : '22px';
    node.parentElement.replaceChild(spinner, node);
  }
});
export default sanitize;
