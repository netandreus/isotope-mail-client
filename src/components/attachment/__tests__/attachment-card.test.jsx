import React from 'react';
import {shallow} from 'enzyme';
import {AttachmentCard} from '..//attachment-card';
import * as messageService from '../../../services/message';

const MOCK_ATTACHMENT = {fileName: 'test-attachment.file', size: 313373};

describe('AttachmentCard component test suite', () => {
  test('Snapshot render, should render attachment card for file', () => {
    const attachmentCard = shallow(<AttachmentCard attachment={MOCK_ATTACHMENT}/>);
    expect(attachmentCard).toMatchSnapshot();
  });
  test('download, not downloading, should trigger download', async () => {
    // Given
    const fetchPromise = Promise.resolve({});
    messageService.downloadAttachment = jest.fn(() => fetchPromise);
    const attachmentCard = shallow(<AttachmentCard attachment={MOCK_ATTACHMENT}/>);

    // When
    attachmentCard.find('.attachment').simulate('click');
    await fetchPromise;
    attachmentCard.update();

    // Then
    expect(messageService.downloadAttachment).toHaveBeenCalledTimes(1);
    expect(attachmentCard.state().downloading).toBe(false);
  });
  test('download, already downloading, should return', async () => {
    // Given
    messageService.downloadAttachment = jest.fn();
    const attachmentCard = shallow(<AttachmentCard attachment={MOCK_ATTACHMENT}/>);
    attachmentCard.setState({downloading: true});
    attachmentCard.update();

    // When
    attachmentCard.find('.attachment').simulate('click');

    // Then
    expect(messageService.downloadAttachment).not.toHaveBeenCalled();
    expect(attachmentCard.state().downloading).toBe(true);
  });
});
