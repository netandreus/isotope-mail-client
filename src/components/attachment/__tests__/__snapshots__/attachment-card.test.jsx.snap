// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentCard component test suite Snapshot render, should render attachment card for file 1`] = `
<div
  className="attachment "
  isotip="test-attachment.file"
  isotip-position="bottom"
  onClick={[Function]}
>
  <Spinner
    canvasClassName="canvas"
    className="spinner"
    pathClassName=""
    visible={false}
  />
  <div
    className="material-icons mdc-list-item__graphic icon"
  >
    attach_file
  </div>
  <div
    className="fileInfo"
  >
    <div
      className="fileName"
    >
      test-attachment.file
    </div>
    <div
      className="size"
    >
      306.03 KiB
    </div>
  </div>
</div>
`;
